[wasm] Improve AOT builds

radical opened the pull request:
- Fix incremental AOT builds:
   If you build with AOT, and then build again, then the build breaks. The main reason is that we cil-strip the runtime 
   pack assemblies in-place, which breaks the subsequent build.

- This also fixes wasm targets to work better with blazor workloads.
- Fix targets to work better on windows
- Add a new `RunWithEmSdkEnv` task

radical committed:
[wasm] pack: Let the workload configure the runtime pack path

radical committed:
[wasm] update webassembly pack description

radical committed:
[wasm] Makefile: fix being able to `make build`

radical committed:
[wasm] Fix AOT compilation to work before publish

Instead of using runtime pack, and other assemblies from the
publish folder, we are now using them directly from the source
location. This means that `System.Private.CoreLib` can be in a different
directory, than the assembly being precompiled.

So, we need to set `MONO_PATH` to the relevant paths that will have the assemblies.

This manifests during blazor workload based builds.

radical committed:
[wasm][aot] Don't strip the source assemblies in place

Currently we run `mono-cil-strip` on assemblies, in the AOT case, but we
do that in-place.

This is problematic because the assemblies from the runtime pack will
get modified in their original location (nuget/etc), since we don't use
them from the publish folder anymore.

- This would mean that a subsequent attempt to build with AOT would fail
  because now the original assemblies are already stripped!

- this also becomes apparent if you have the runtime pack installed in a
  system dotnet folder, eg. as a pack - `/usr/local/share/dotnet/packs/Microsoft.NETCore.App.Runtime.browser-wasm/`,
  because a regular user wouldn't have permissions to modify that.

Solution: Write the stripped assemblies into the obj dir

radical committed:
[wasm] Make the final list of wasm assemblies as an item

- Blazor skips the generation of the app bundle when running
 `WasmBuildApp` target.
- Because of the previous commit, the locations for the final assemblies
  could be different (when aot+stripping), which needs to be surfaced
  for blazor. So, generate a `@(WasmAssembliesFinal)` item for that.

radical committed:
[wasm] Allow enabling assert messages with $(MonoEnableAssertMessages)

radical committed:
[wasm][aot] WasmApp.targets: check that we have .bc file for each assembly

Just a sanity check to ensure that due to some build authoring error, we
are not building dotnet.js/wasm with no .bc files! That just shows up as
an assert:

`* Assertion at /Users/radical/dev/r3/src/mono/mono/mini/aot-runtime.c:2651, condition `<disabled>' not met`

.. which is:
    `g_assert (info->version == MONO_AOT_FILE_VERSION);`

..  `0 != MONO_AOT_FILE_VERSION`

radical committed:
[wasm] remove UsingTask for non-exitant WasmAOT

radical committed:
MonoCrossAot pkg: use mono-aot-cross.exe on windows

radical committed:
[wasm] Fix up emscripten command invocations for windows

radical committed:
[wasm] WasmApp.targets: move some property inits closer to their use

radical committed:
[wasm] fix build for browser-profile sample

radical committed:
[wasm] fix emcc invocation for getting the version

pranavkm commented on src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.MonoCrossAOT.Sdk.props:
```diff
@@ -1,5 +1,6 @@
 <Project>
   <PropertyGroup>
     <MonoAotCrossCompilerPath>$(MSBuildThisFileDirectory)..\tools\mono-aot-cross</MonoAotCrossCompilerPath>
+    <MonoAotCrossCompilerPath Condition="'$(TargetOS)' == 'windows'">$(MonoAotCrossCompilerPath).exe</MonoAotCrossCompilerPath>
```
```suggestion
    <MonoAotCrossCompilerPath Condition="$([MSBuild]::IsOsPlatform('Windows'))">$(MonoAotCrossCompilerPath).exe</MonoAotCrossCompilerPath>
```

pranavkm commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -86,7 +81,11 @@
       <_WasmAssemblies Remove="@(_WasmAssemblies)" />
+
+      <_WasmAOTSearchPaths Include="$(MicrosoftNetCoreAppRuntimePackRidDir)\native" />
+      <_WasmAOTSearchPaths Include="$(MicrosoftNetCoreAppRuntimePackRidDir)\lib\net6.0" />
```
Are you able to use `$(TargetFramework)` here instead? Avoids having to keep this file updated between major releases.

radical replied:
`$(TargetFramework)` can be different from what the runtime pack has, right? I'm not sure how I can get the one to be used with the pack.

pranavkm replied:
Not sure I follow. The TFM for the runtime pack is determined by the project's TFM (by the SDK). I haven't seen a case where they would be different.

pranavkm commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -207,15 +229,25 @@
+    </PropertyGroup>
+
+    <PropertyGroup>
+      <_EmsdkEnvSourceCommand Condition="$([MSBuild]::IsOSPlatform('WINDOWS'))">$(EMSDK_PATH)\emsdk_env.bat</_EmsdkEnvSourceCommand>
```
```suggestion
      <_EmsdkEnvSourceCommand Condition="$([MSBuild]::IsOSPlatform('WINDOWS'))">$(EMSDK_PATH)\emsdk_env.bat &gt; nul 2&gt;&amp;1 </_EmsdkEnvSourceCommand>
```

pranavkm commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -207,15 +229,25 @@
 
-    <Exec Command="bash -c '$(_EmccCommand) --version | head -1 > emcc-version.txt'" IgnoreStandardErrorWarningFormat="true" WorkingDirectory="$(_WasmIntermediateOutputPath)" />
+    <Exec Condition="!$([MSBuild]::IsOSPlatform('WINDOWS'))" IgnoreStandardErrorWarningFormat="true" WorkingDirectory="$(_WasmIntermediateOutputPath)"
+          Command="bash -c '$(_EmccCommand) --version | head -1 > emcc-version.txt'" />
```
Could you file a tracking issue to determine if shelling out to bash could be removed? It adds an extra dependency that might not be available in all environments.

radical replied:
On non-windows platforms, we would it, I believe, to run `emsdk_env.sh`. What platforms might have it missing, and we would be able to install Emscripten?

pranavkm commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -207,15 +229,25 @@
+    <Exec Condition="!$([MSBuild]::IsOSPlatform('WINDOWS'))" IgnoreStandardErrorWarningFormat="true" WorkingDirectory="$(_WasmIntermediateOutputPath)"
+          Command="bash -c '$(_EmccCommand) --version | head -1 > emcc-version.txt'" />
+
+    <Exec Condition="$([MSBuild]::IsOSPlatform('WINDOWS'))" IgnoreStandardErrorWarningFormat="true" WorkingDirectory="$(_WasmIntermediateOutputPath)"
```
You could use ConsoleToMsBuild. https://stackoverflow.com/a/16804205

pranavkm commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -318,4 +356,10 @@ EMSCRIPTEN_KEEPALIVE void mono_wasm_load_profiler_aot (const char *desc) { mono_
+
+  <Target Name="_AfterWasmBuildApp">
+    <ItemGroup>
+      <WasmAssembliesFinal Include="@(_WasmAssemblies)" />
```
Should the SDK start using this item group?

radical replied:
Yeah, once we merge this

pranavkm commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -257,20 +289,26 @@
-    <Exec Command="bash -c '$(_EmccCommand) $(EmccLDFlags) --js-library $(_WasmSrcDir)/library_mono.js --js-library $(_WasmSrcDir)/binding_support.js --js-library $(_WasmSrcDir)/dotnet_support.js --js-library $(_WasmSrcDir)/pal_random.js @(_WasmAssemblies->'%(LlvmBitcodeFile)', ' ') @(_WasmObjects, ' ') -o $(_WasmIntermediateOutputPath)dotnet.js'" IgnoreStandardErrorWarningFormat="true" StandardO ...
-    <Exec Condition="'$(WasmNativeStrip)' == 'true'" Command="bash -c '$(EMSDK_PATH)/upstream/bin/wasm-opt --strip-dwarf $(_WasmIntermediateOutputPath)dotnet.wasm -o $(_WasmIntermediateOutputPath)dotnet.wasm'" IgnoreStandardErrorWarningFormat="true" />
+    <Exec Command="$(_EmccCommand) $(EmccCFlags) $(_WasmSrcDir)/driver.c -c -o $(_WasmIntermediateOutputPath)driver.o" IgnoreStandardErrorWarningFormat="true" StandardOutputImportance="Low" />
+    <Exec Command="$(_EmccCommand) $(EmccCFlags) $(_WasmSrcDir)/corebindings.c -c -o $(_WasmIntermediateOutputPath)corebindings.o" IgnoreStandardErrorWarningFormat="true" StandardOutputImportance="Low" />
```
Does it work without shelling out?

radical replied:
Maybe.. I can run `$EMSDK_PATH/emsdk emcc`, but I don't know what it might require to compile this or other stuff. If somebody familiar with emsdk can say that it's safe to use that, then I can do that. Maybe @kg, @vargaz, or @lewing?

radical committed:
[wasm] Add new RunWithEmSdkEnv task

radical committed:
[wasm] Use `RunWithEmSdkEnv` to run `emcc` commands

pranavkm commented on src/tasks/WasmAppBuilder/RunWithEmSdkEnv.cs:
```diff
@@ -0,0 +1,51 @@
+                Command = $"bash -c 'source {envScriptPath} > /dev/null 2>&1 && {Command}'";
+            }
+
+            return base.Execute();
```
re
```suggestion
            return !Log.HasLoggedErrors;
```

pranavkm commented on src/tasks/WasmAppBuilder/RunWithEmSdkEnv.cs:
```diff
@@ -0,0 +1,51 @@
+
+namespace Microsoft.WebAssembly.Build.Tasks
+{
+    public class RunWithEmSdkEnv : Exec
```
👏🏽

radical committed:
[wasm] AotCompiler: use platform specific path separator

radical committed:
[wasm] RunWithEmSdkEnv: quote path on windows

radical committed:
[wasm] Use new ReadEmccVersion to get the versions

This simplifies the target too, and works on windows.

radical committed:
Update src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.MonoCrossAOT.Sdk.props

Co-authored-by: Pranav K <prkrishn@hotmail.com>

radical committed:
fix messageimportance for debug msg

radical committed:
Fix emcc invocation on windows, thanks to @pranavkm

radical committed:
Merge remote-tracking branch 'origin/master'

kg commented on src/tasks/WasmAppBuilder/ReadEmccVersion.cs:
```diff
@@ -0,0 +1,35 @@
+                return false;
+            }
+
+            Version = File.ReadLines(VersionFile).First();
```
I don't think we need a special task for this. MSBuild has one built in:
https://docs.microsoft.com/en-us/visualstudio/msbuild/readlinesfromfile-task?view=vs-2019

lewing replied:
yeah you should be able to read the version with something like
```
    <ReadLinesFromFile File="@(VersionFile)">
      <Output TaskParameter="Lines" ItemName="VersionLines" />
    </ReadLinesFromFile>
    <PropertyGroup>
      <EmccVersion>@(VersionLines)</EmccVersion>
    </PropertyGroup>
```

radical replied:
That's what is already there. The problem was reading one line, on windows. The whole output is like:
```
emcc (Emscripten gcc/clang-like replacement) 2.0.11 (6e28e4fa4fa1bc50d58b9ddbbb9603a3cf21ea9e)
Copyright (C) 2014 the Emscripten authors (see AUTHORS.txt)
This is free and open source software under the MIT license.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

```

pranavkm replied:
Could you use `StartsWith` instead?

```xml
<Error Condition="!$(ActualVersion.StartsWith($(ExpectedVersion))" Text="..." />
```

that solves this issue.

pranavkm replied:
Unrelated, how willing would emsdk be to change this command output? It's really weird that asking for a version includes licensing information.

marek-safar replied:
We could submit PR

vargaz replied:
The output is probably not designed to be machine parsable. Can't msbuild read the first line only, or extract the first item in @(VersionLines) ?

pranavkm replied:
Something like this might work:

```xml
<EmccVersion>@(VersionLines->Reverse())</EmccVersion>
```

radical replied:
`<EmccVersion Condition="$([System.String]::new('%(Lines.Identity)').Contains('emcc'))">@(Lines)</EmccVersion>`

lewing replied:
try `"$([System.IO.File]::ReadAllText('$(VersionsFile)')[0])"`

radical replied:
It doesn't let me call `File.ReadAllLines`. So, now I have:
```xml
    <ItemGroup>
      <_ReversedVersionLines Include="@(_VersionLines->Reverse())" />
    </ItemGroup>
    <PropertyGroup>
      <ActualEmccVersion>%(_ReversedVersionLines.Identity)</ActualEmccVersion>
    </PropertyGroup>
```

I think this is a decent enough solution for now, and it doesn't depend on any specific string/char in the output.

lewing replied:
I meant ReadAllLines but I see that isn't available :(
The Reverse() call looks best

radical committed:
Don't strip assemblies with mono-cil-strip for aotinterp mode

radical committed:
[wasm] Remove ReadEmccVersion task, and use msbuild to get the version

.. string instead. We do this by looking for the line with `'emcc'` in
it, since the other lines are the license text. Hopefully, this isn't
too brittle, since we are invoking `emcc --version`.

radical committed:
Fix emcc version extraction to not depend on a specific string

lewing reviewed (approved):
This looks good to me, I talked to @radical about some improvements in allowing InterpOnly metadata on the assembly items in AotIterp mode and a couple of naming changes in a follow up but we want to unblock @pranavkm for now.
